src/mjolnir/landmarks.cc

```diff
@@ -296,6 +318,226 @@ bool BuildLandmarkFromPBF(const boost::property_tree::ptree& pt,
+                 const std::string& tile_dir,
+                 const std::string& db_name,
+                 const size_t& thread_number,
+                 const size_t& total_threads,
+                 std::promise<std::tuple<size_t, size_t, size_t>>& stats) {
+  // open the database and initialize a unique pointer to graph tile builder
+  std::unique_ptr<GraphTileBuilder> tile_builder_ptr = nullptr;
+
+  // const std::string dbname = pt.get_child("mjolnir").get<std::string>("landmarks", "");
+  LandmarkDatabase db(db_name, true);
+
+  // stats
+  size_t updated_tiles = 0, updated_edges = 0, updated_landmarks = 0;
+  GraphId last_edge, last_tile;
+
+  // modulus count to assign tiles to threads
+  size_t tile_count = static_cast<size_t>(-1);
+
+  for (auto it = seq_file.begin(); it != seq_file.end(); ++it) {
+    if ((*it).first.Tile_Base() != last_tile) {
+      last_tile = (*it).first.Tile_Base();
+      tile_count++;
+    }
+    if (tile_count % total_threads != thread_number) {
+      continue;
+    }
+
+    if (!tile_builder_ptr ||
+        tile_builder_ptr->header_builder().graphid().Tile_Base() != (*it).first.Tile_Base()) {
+      // store the previously updated tile
+      if (tile_builder_ptr) {
+        tile_builder_ptr->StoreTileData();
+        updated_tiles++;
+      }
+      // reset the tile builder ptr to the current tile
+      tile_builder_ptr.reset(new GraphTileBuilder(tile_dir, (*it).first.Tile_Base(), true));
+    }
+
+    // retrieve the landmark to be added
+    const std::vector<Landmark> landmark =
+        db.get_landmarks_by_ids({static_cast<int64_t>((*it).second)});
+    if (landmark.size() != 1) {
+      throw std::logic_error("Incorrect result size " + std::to_string(landmark.size()) +
+                             " of retrieved landmarks, which should be 1");
+    }
+    // add the landmark to the tile
+    GraphId edge_id = (*it).first;
+    tile_builder_ptr->AddLandmark(edge_id, landmark[0]);
+
+    // update the counting numbers
+    updated_landmarks++;
+    if (last_edge != (*it).first) {
+      updated_edges++;
+      last_edge = (*it).first;
+    }
+  }
+  // store the last updated tile
+  if (tile_builder_ptr) {
+    tile_builder_ptr->StoreTileData();
+    updated_tiles++;
+  }
+
+  // set the stats
+  stats.set_value(std::make_tuple(updated_tiles, updated_edges, updated_landmarks));
+}
+
+// Add all landmarks to tiles
+bool AddLandmarks(const boost::property_tree::ptree& pt) {
+  LOG_INFO("Starting adding landmarks to tiles...");
+
+  const size_t num_threads =
+      pt.get<size_t>("mjolnir.concurrency", std::thread::hardware_concurrency());
+  const std::string db_name = pt.get_child("mjolnir").get<std::string>("landmarks_db", "");
+
+  // get tile access
+  baldr::GraphReader reader(pt.get_child("mjolnir"));
+
+  // get all tile ids and sort the tiles in descending order by size to balance the threads
+  auto tileset = reader.GetTileSet(1);
```

Comment:
it is possible in a global tileset that we have coverage only at level 2 for some places and we'd still like to get landmarks there. we'll probably need to fix this...